chain-factory/src/factory.rs

```diff
@@ -30,29 +36,105 @@ pub trait FactoryModule {
         min_validators: usize,
         max_validators: usize,
         min_stake: BigUint,
+        chain_name: ManagedBuffer,
         additional_stake_required: MultiValueEncoded<StakeMultiArg<Self::Api>>,
     ) {
         let payment_amount = self.call_value().egld_value().clone_value();
         let deploy_cost = self.deploy_cost().get();
         require!(payment_amount == deploy_cost, "Invalid payment amount");
 
-        let caller = self.blockchain().get_caller();
         let source_address = self.chain_config_template().get();
+        let metadata =
+            CodeMetadata::PAYABLE_BY_SC | CodeMetadata::UPGRADEABLE | CodeMetadata::READABLE;
+        let args = self.get_deploy_chain_config_args(
+            &min_validators,
+            &max_validators,
+            &min_stake,
+            &additional_stake_required,
+        );
+
+        let sc_address = self
+            .tx()
+            .raw_deploy()
+            .gas(self.blockchain().get_gas_left())
+            .from_source(source_address)
+            .code_metadata(metadata)
+            .arguments_raw(args)
+            .returns(ReturnsNewManagedAddress)
+            .sync_call();
+
+        self.all_deployed_contracts(ManagedBuffer::new())
+            .insert(sc_address);
+
+        let chain_id = self.generate_chain_id();
+        let chain_info = ChainInfo {
+            name: chain_name,
+            chain_id,
+        };
+        self.chain_info().set(chain_info);
+    }
+
+    #[only_owner]
+    #[endpoint(deployHeaderVerifier)]
+    fn deploy_header_verifier(
+        &self,
+        chain_id: ManagedBuffer,
+        bls_pub_keys: MultiValueEncoded<ManagedBuffer>,
+    ) {
+        let source_address = self.header_verifier_template().get();
+        let metadata =
+            CodeMetadata::PAYABLE_BY_SC | CodeMetadata::UPGRADEABLE | CodeMetadata::READABLE;
+
+        let mut args = ManagedArgBuffer::new();
+        args.push_multi_arg(&bls_pub_keys);
+
+        let header_verifier_address = self
+            .tx()
+            .raw_deploy()
+            .gas(self.blockchain().get_gas_left())
+            .from_source(source_address)
+            .code_metadata(metadata)
+            .arguments_raw(args)
+            .returns(ReturnsNewManagedAddress)
+            .sync_call();
+
+        self.all_deployed_contracts(chain_id)
+            .insert(header_verifier_address);
+    }
+
+    #[only_owner]
+    #[endpoint(deployCrossChainOperation)]
+    fn deploy_cross_chain_operation(
+        &self,
+        chain_id: ManagedBuffer,
+        is_sovereign_chain: bool,
+        token_handler_address: ManagedAddress,
+        opt_wegld_identifier: Option<TokenIdentifier>,
+        opt_sov_token_prefix: Option<ManagedBuffer>,
+    ) {
+        let source_address = self.cross_chain_operations_template().get();
+
         let metadata =
```

Comment:
Why not read the metadata from the source?

Reply:
do you mean to create storage mappers for those contracts metadata ?

Reply:
You can use `self.blockchain().get_code_metadata(&template_address)` to get the metadata from a template contract.
Here is an example: https://github.com/multiversx/mx-contracts-rs/blob/8b51d5fa6fdce4aa35abb294dd3e3b172db1246d/contracts/proxy-deployer/src/contract_interactions.rs#L45